firedrake/slate/slate.py

```diff
@@ -0,0 +1,767 @@
+"""SLATE is a symbolic language defining a framework for
+performing linear algebra operations on finite element
+tensors. It is similar in principle to most linear algebra
+libraries in notation.
+
+The design of SLATE was heavily influenced by UFL, and
+utilizes much of UFLs functionality for FEM-specific form
+manipulation.
+
+Unlike UFL, however, once forms are assembled into SLATE
+`Tensor` objects, one can utilize the operations defined
+in SLATE to express complicated linear algebra operations.
+(Such as the Schur-complement reduction of a block-matrix
+system)
+
+All SLATE expressions are handled by a specialized linear
+algebra compiler, the SLATE Linear Algebra Compiler (SLAC),
+which interprets SLATE expressions and produces C++ kernel
+functions to be executed within the Firedrake architecture.
+
+Written by: Thomas Gibson (t.gibson15@imperial.ac.uk)
+"""
+
+import hashlib
+import operator
+
+from slate_assertions import *
+from slate_equation import SlateEquation
+
+from ufl.algorithms.domain_analysis import canonicalize_metadata
+from ufl.algorithms.map_integrands import map_integrand_dags
+from ufl.algorithms.multifunction import MultiFunction
+from ufl.algorithms.signature import compute_expression_hashdata, compute_terminal_hashdata
+from ufl.domain import join_domains, sort_domains
+from ufl.form import _sorted_integrals
+
+
+__all__ = ['CheckRestrictions', 'RemoveRestrictions',
+           'SlateIntegral', 'Tensor',
```

Comment:
Given that these are imported into the global namespace, I think `CheckRestrictions` and `RemoveRestrictions` (and maybe `SlateIntegral`) shouldn't appear on this list.  Moreover `RemoveRestrictions` is really part of the compiler, not the language, so should it move across?